vllm/compilation/cuda_graph.py

```diff
@@ -0,0 +1,169 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+
+import dataclasses
+from contextlib import ExitStack
+from typing import Any, Callable, Optional
+from unittest.mock import patch
+
+import torch
+
+import vllm.envs as envs
+from vllm.compilation.counter import compilation_counter
+from vllm.config import CUDAGraphRuntimeStyle, VllmConfig
+from vllm.forward_context import get_forward_context
+from vllm.logger import init_logger
+from vllm.utils import weak_ref_tensors
+from vllm.platforms import current_platform
+
+logger = init_logger(__name__)
+
+
+@dataclasses.dataclass
+class CUDAGraphEntry:
+    runtime_shape: int
+    cudagraph: Optional[torch.cuda.CUDAGraph] = None
+    output: Optional[Any] = None
+
+    # for cudagraph debugging, track the input addresses
+    # during capture, and check if they are the same during replay
+    input_addresses: Optional[list[int]] = None  
+
+
+@dataclasses.dataclass
+class CUDAGraphOptions:
+    debug_log_enable: bool = True
+    gc_disable: bool = False
+    weak_ref_output: bool = True
+    usage_str: Optional[str] = None # For debug logging only
+
+
+class CUDAGraphWrapper:
+    """
+    This class simply wrap a runnable for cudagraph functionality,
+    taking responsibility of capturing cudagraph and running the replay.
+    """
+
+    def __init__(self,
+                 runnable: Callable,
+                 vllm_config: VllmConfig,
+                 runtime_style: CUDAGraphRuntimeStyle,
+                 graph_pool: Any = current_platform.get_global_graph_pool(),
```

Comment:
Sorry for this suggestion, didn't realize defaults don't get reevaluated for every call 😅